contracts/ERC721Collateralizer.sol

```diff
@@ -0,0 +1,426 @@
+
+        // Ensure that this contract is in fact the owner of the collateral.
+        require(erc721token.ownerOf(collateralTokenID) == custodian);
+
+        // Store debtor in mapping, effectively demarcating that the agreement is now collateralized.
+        agreementToDebtor[agreementId] = debtor;
+
+        // Emit an event that collateral has been secured.
+        CollateralLocked(agreementId, collateralTokenAddress, collateralTokenID);
+
+        return true;
+    }
+
+    /**
+     * Returns collateral to the debt agreement's debtor if and only if the debt agreement's
+     * term has lapsed and the total expected repayment value has been repaid.
+     *
+     * @param agreementId bytes32 The debt agreement's ID
+     */
+    function returnCollateral(
+        bytes32 agreementId
+    )
+        public
+        whenNotPaused
+    {
+        // The address of the ERC721 contract.
+        address collateralTokenAddress;
+        // The token id, associating the collateral token and the given ERC721 contract.
+        uint256 collateralTokenID;
+        // The terms contract according to which this asset is being collateralized.
+        TermsContract termsContract;
+
+        // Fetch all relevant collateralization parameters.
+        (
+            collateralTokenAddress,
+            collateralTokenID,
+            termsContract
+        ) = retrieveCollateralParameters(agreementId);
+
+        // Withdrawal can only occur if the collateral has yet to be withdrawn.
+        // When we withdraw collateral, we reset the collateral agreement
+        // in a gas-efficient manner by resetting the address of the debtor to 0.
+        require(agreementToDebtor[agreementId] != address(0));
+
+        // Determine debtor of the collateral.
+        address debtor = agreementToDebtor[agreementId];
+
+        // Mark agreement's collateral as withdrawn by setting the agreement's debtor to 0x0.
+        delete agreementToDebtor[agreementId];
+
+        // Ensure that the debt is not in a state of default.
+        require(
+            termsContract.getExpectedRepaymentValue(
+                agreementId,
+                termsContract.getTermEndTimestamp(agreementId)
+            ) <= termsContract.getValueRepaidToDate(agreementId)
+        );
+
+        // Instantiate an instance of the token, in order to initiate a transfer.
+        ERC721 erc721token = ERC721(collateralTokenAddress);
+
+        // Apply a hack for CryptoKitties, granting transfer approval to this contract itself.
+        if (cryptoKittiesContract == collateralTokenAddress) {
+            erc721token.approve(address(this), collateralTokenID);
+        }
+
+        // Transfer the collateral this contract was holding in escrow back to the debtor.
+        erc721token.transferFrom(address(this), debtor, collateralTokenID);
+
+        // Ensure that transfer has succeeded.
+        require(erc721token.ownerOf(collateralTokenID) == debtor);
```

Comment:
These sorts of post-call invariant-checks that should only get triggered if a 721 is mis-implemented are a great idea, but I think it's generally considered a better practice to use `assert` statements for them.  See: https://ethereum.stackexchange.com/questions/15166/difference-between-require-and-assert-and-the-difference-between-revert-and-thro